File: leetCode/graph/Flood_Fill_733/Main.java
class Solution {
    public void dfs(int[][] image, int sr, int sc, int newColor, int prevColor) {
        if(sr < 0 || sr >= image.length || sc < 0 || sc >= image[0].length) return;
        if(image[sr][sc] == newColor || image[sr][sc] != prevColor) return;

        image[sr][sc] = newColor;
        dfs(image, sr + 1, sc, newColor, prevColor);
        dfs(image, sr - 1, sc, newColor, prevColor);
        dfs(image, sr, sc + 1, newColor, prevColor);
        dfs(image, sr, sc - 1, newColor, prevColor);
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        dfs(image, sr, sc, newColor, image[sr][sc]);
        return image;
    }
}

File: leetCode/graph/Flood_Fill_733/main.py
class Solution:
    def dfs(self, image, x, y, new_c, prev_c):
        if not self.is_inside(x, y):
            return

        if image[x][y] != prev_c or image[x][y] == new_c:
            return

        image[x][y] = new_c
        self.dfs(image, x + 1, y, new_c, prev_c)
        self.dfs(image, x - 1, y, new_c, prev_c)
        self.dfs(image, x, y + 1, new_c, prev_c)
        self.dfs(image, x, y - 1, new_c, prev_c)

    def floodFill(self, image, sr, sc, newColor):
        m, n = len(image), len(image[0])

        self.is_inside = lambda x, y: 0 <= x < m and 0 <= y < n

        self.dfs(image, sr, sc, newColor, image[sr][sc])

        return image


a = Solution()
a.floodFill([[1,1,1],[1,1,0],[1,0,1]], 1, 1, 2)
pass